File: examples/python/wordcount/wordcount.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Flink WordCount Python示例

功能：统计文本流中的单词频次

运行方式：
1. 本地运行: python wordcount.py
2. 指定输入: python wordcount.py --input data.txt
3. 集群提交: flink run -py wordcount.py

依赖安装：
    pip install -r requirements.txt

作者: AnalysisDataFlow Project
版本: 1.0.0
"""

import argparse
import logging
from typing import Iterable

from pyflink.common import Types
from pyflink.datastream import StreamExecutionEnvironment, TimeCharacteristic
from pyflink.datastream.functions import FlatMapFunction, RuntimeContext
from pyflink.datastream.window import TumblingProcessingTimeWindows, Time

# 配置日志
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(levelname)s - %(message)s'
)
logger = logging.getLogger(__name__)


class Tokenizer(FlatMapFunction):
    """
    分词器：将文本行拆分为单词
    
    输出格式: (word, 1)
    """
    
    def flat_map(self, value: str, collector):
        """
        将输入行分割为单词并输出(word, count)元组
        
        Args:
            value: 输入文本行
            collector: 结果收集器
        """
        # 转换为小写并按非单词字符分割
        words = value.lower().split()
        
        for word in words:
            # 去除标点符号
            clean_word = ''.join(char for char in word if char.isalnum())
            if clean_word:
                collector.collect((clean_word, 1))


def create_demo_data() -> list:
    """
    创建演示数据
    
    Returns:
        示例文本列表
    """
    return [
        "Hello Flink",
        "Hello World",
        "Flink is awesome",
        "Stream processing with Flink",
        "Hello Flink World",
        "Real-time analytics",
        "Flink rocks",
        "Hello again",
        "Apache Flink is great",
        "Stream processing made easy"
    ]


def run_wordcount(input_file: str = None):
    """
    运行WordCount程序
    
    Args:
        input_file: 可选的输入文件路径，不指定则使用演示数据
    """
    logger.info("=" * 50)
    logger.info("Flink WordCount Python Example")
    logger.info("=" * 50)
    
    # 创建执行环境
    env = StreamExecutionEnvironment.get_execution_environment()
    env.set_parallelism(2)
    
    # 设置时间语义（处理时间）
    env.set_stream_time_characteristic(TimeCharacteristic.ProcessingTime)
    
    # 创建数据源
    if input_file:
        logger.info(f"从文件读取数据: {input_file}")
        # 从文件读取
        data_stream = env.read_text_file(input_file)
    else:
        logger.info("使用内置演示数据")
        logger.info("如需从文件读取，请使用: python wordcount.py --input your_file.txt")
        # 使用演示数据
        data = create_demo_data()
        data_stream = env.from_collection(data)
    
    # 数据流处理
    word_counts = (
        data_stream
        # 1. 分词
        .flat_map(
            Tokenizer(),
            output_type=Types.TUPLE([Types.STRING(), Types.INT()])
        )
        # 2. 按单词分组
        .key_by(lambda x: x[0])
        # 3. 5秒滚动窗口
        .window(TumblingProcessingTimeWindows.of(Time.seconds(5)))
        # 4. 求和
        .sum(1)
    )
    
    # 输出结果
    word_counts.print()
    
    # 执行
    logger.info("启动Flink作业...")
    env.execute("Flink WordCount Python")
    logger.info("作业执行完成")


def main():
    """主入口函数"""
    parser = argparse.ArgumentParser(
        description='Flink WordCount Python Example',
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog="""
示例:
  python wordcount.py                    # 使用演示数据运行
  python wordcount.py --input data.txt   # 从文件读取数据
  flink run -py wordcount.py             # 提交到Flink集群
        """
    )
    
    parser.add_argument(
        '--input', '-i',
        type=str,
        help='输入文件路径 (默认使用演示数据)'
    )
    
    parser.add_argument(
        '--parallelism', '-p',
        type=int,
        default=2,
        help='并行度 (默认: 2)'
    )
    
    args = parser.parse_args()
    
    try:
        run_wordcount(args.input)
    except Exception as e:
        logger.error(f"程序执行出错: {e}")
        raise


if __name__ == '__main__':
    main()

File: release/v3.0.0/docs/examples/java/wordcount/src/main/java/com/example/flink/WordCount.java
package com.example.flink;

import org.apache.flink.api.common.eventtime.WatermarkStrategy;
import org.apache.flink.api.common.functions.FlatMapFunction;
import org.apache.flink.api.java.tuple.Tuple2;
import org.apache.flink.streaming.api.datastream.DataStream;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.streaming.api.windowing.assigners.TumblingProcessingTimeWindows;
import org.apache.flink.streaming.api.windowing.time.Time;
import org.apache.flink.util.Collector;

/**
 * Flink WordCount 示例程序
 * 
 * 功能：从socket或内置数据流读取文本，统计单词频次
 * 
 * 运行方式：
 * 1. 本地运行: mvn compile exec:java -Dexec.mainClass="com.example.flink.WordCount" -Plocal
 * 2. 打包提交: mvn clean package -Plocal && flink run target/flink-wordcount-1.0.0.jar
 * 
 * @author AnalysisDataFlow Project
 * @version 1.0.0
 */
public class WordCount {

    public static void main(String[] args) throws Exception {
        // 创建执行环境
        final StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
        
        // 设置并行度
        env.setParallelism(2);
        
        // 创建数据源 - 使用内置数据流便于测试
        DataStream<String> source;
        
        if (args.length > 0 && "socket".equals(args[0])) {
            // 从socket读取 (需要先运行: nc -lk 9999)
            String hostname = args.length > 1 ? args[1] : "localhost";
            int port = args.length > 2 ? Integer.parseInt(args[2]) : 9999;
            System.out.println("从Socket读取数据: " + hostname + ":" + port);
            System.out.println("请先运行: nc -lk " + port);
            source = env.socketTextStream(hostname, port);
        } else {
            // 使用内置数据流进行演示
            System.out.println("使用内置数据流进行演示");
            System.out.println("如需从Socket读取，请运行: java -jar target/flink-wordcount-1.0.0.jar socket localhost 9999");
            source = env.fromData(
                "Hello Flink",
                "Hello World",
                "Flink is awesome",
                "Stream processing with Flink",
                "Hello Flink World",
                "Real-time analytics",
                "Flink rocks",
                "Hello again"
            );
        }

        // 数据流处理链
        DataStream<Tuple2<String, Integer>> wordCounts = source
            // 1. 分词处理
            .flatMap(new Tokenizer())
            // 2. 分配水印（允许2秒延迟）
            .assignTimestampsAndWatermarks(
                WatermarkStrategy.<Tuple2<String, Integer>>forMonotonousTimestamps()
                    .withIdleness(java.time.Duration.ofSeconds(2))
            )
            // 3. 按单词分组
            .keyBy(value -> value.f0)
            // 4. 5秒滚动窗口
            .window(TumblingProcessingTimeWindows.of(Time.seconds(5)))
            // 5. 求和
            .sum(1);

        // 输出结果到控制台
        wordCounts.print().setParallelism(1);

        // 执行程序
        System.out.println("=================================");
        System.out.println("WordCount程序已启动");
        System.out.println("=================================");
        env.execute("Flink WordCount Example");
    }

    /**
     * 分词器：将文本行拆分为单词
     */
    public static class Tokenizer implements FlatMapFunction<String, Tuple2<String, Integer>> {
        
        @Override
        public void flatMap(String value, Collector<Tuple2<String, Integer>> out) {
            // 转换为小写并按非单词字符分割
            String[] words = value.toLowerCase().split("\\W+");
            
            for (String word : words) {
                if (word.length() > 0) {
                    out.collect(new Tuple2<>(word, 1));
                }
            }
        }
    }
}
